docs/zkapps/o1js/indexed-merkle-map.mdx

```diff
@@ -0,0 +1,263 @@
+    indexedMerkleMap: IndexedMerkleMap31
+  ) {
+    // Validate the integrity of the Merkle Map input
+    const currentRoot = this.mapRoot.getAndRequireEquals();
+    currentRoot.assertEquals(
+      indexedMerkleMap.root,
+      'Off-chain Indexed Merkle Map is out of sync!'
+    );
+
+    // Insert a new key-value pair
+    indexedMerkleMap = indexedMerkleMap.clone();
+    indexedMerkleMap.insert(newKey, newValue);
+
+    // Update the on-chain map root
+    const newMapRoot = indexedMerkleMap.root;
+    this.mapRoot.set(newMapRoot);
+  }
+
+  @method async updateExistingLeaf(
+    key: Field,
+    newValue: Field,
+    indexedMerkleMap: IndexedMerkleMap31
+  ) {
+    // Validate integrity of the Merkle Map input
+    const currentRoot = this.mapRoot.getAndRequireEquals();
+    currentRoot.assertEquals(
+      indexedMerkleMap.root,
+      'Off-chain Indexed Merkle Map is out of sync!'
+    );
+
+    indexedMerkleMap = indexedMerkleMap.clone();
+
+    /**
+     * Proves that the key exists
+     * Updates an existing leaf
+     * Returns the previous value
+     */
+    indexedMerkleMap.update(key, newValue);
+
+    // Update the on-chain map root
+    const newMapRoot = indexedMerkleMap.root;
+    this.mapRoot.set(newMapRoot);
+  }
+}
+```
+
+:warning: Direct modification of a method input can lead to errors. To avoid this, perform changes on a cloned Merkle Map rather than altering the map passed as a method input.
+
+### Interacting with a smart contract utilizing an indexed Merkle Map
+
+To interact with a zkapp that utilizes an `IndexedMerkleMap`, instantiate a map instance and pass it as an argument to the contract’s method within a Mina transaction.
+
+In addition, ensure you synchronize your off-chain map with on-chain updates to maintain data integrity and prepare for subsequent interactions.
+
+```ts
+// Instantiate a new Indexed Merkle Map
+indexedMerkleMap = new IndexedMerkleMap31();
+
+const insertTx = await Mina.transaction(userPubKey, async () => {
+  await contract.insertNewLeaf(Field(1), Field(1234), indexedMerkleMap);
+});
+
+await insertTx.prove();
+await insertTx.sign([userKey]).send();
+
+// Synchornize the off-chain Indexed Merkle Map to match the on-chain state root
+indexedMerkleMap.insert(Field(1), Field(1234));
+
+console.log(
+  indexedMerkleMap.root.toBigInt() === contract.mapRoot.get().toBigInt()
+);
+console.log(indexedMerkleMap.data.get().sortedLeaves);
+
+const updateTx = await Mina.transaction(userPubKey, async () => {
+  await contract.updateExistingLeaf(Field(1), Field(5678), indexedMerkleMap);
+});
+
+await updateTx.prove();
+await updateTx.sign([userKey]).send();
+
+// Synchronize the off-chain Indexed Merkle Map to match the on-chain state root
+indexedMerkleMap.update(Field(1), Field(5678));
+
+console.log(
+  indexedMerkleMap.root.toBigInt() === contract.mapRoot.get().toBigInt()
+);
+console.log(indexedMerkleMap.data.get().sortedLeaves);
+```
+
+## Indexed Merkle Map - API reference
```

Comment:
I feel strongly that this should not be included here.  This definitely belongs in the Typedoc reference, so we should work on generating that correctly instead of manually adding auto-generated type declarations here.